src/Hosting/TestHost/src/ClientHandler.cs

```diff
@@ -65,7 +66,39 @@ internal ClientHandler(PathString pathBase, ApplicationWrapper application)
             var contextBuilder = new HttpContextBuilder(_application, AllowSynchronousIO, PreserveExecutionContext);
 
             var requestContent = request.Content ?? new StreamContent(Stream.Null);
-            var body = await requestContent.ReadAsStreamAsync();
+
+            var requestPipe = new Pipe();
+
+            var copyRequestTask = Task.Run(async () =>
```

Comment:
How do you cancel this loop?

Reply:
It's easy to cancel this, the question is when should it be cancelled?

Reply:
HttpContent.CopyToAsync doesn't support CancellationToken. It is being added in 5.0. I'll add a comment here linking to the issue.

Reply:
The workaround is a wrapping stream that does support cancellation

Reply:
Ehhh, do you want to do this now? `ReadAsStreamAsync` that this is replacing doesn't support cancellation either. I'd rather leave this as is and then add a CT when https://github.com/dotnet/corefx/issues/9071 is done in 5.0.

Reply:
Only if we need to cancel that’s why I asked when do we need to cancel.

Reply:
ReadAsStreamAsync wasn't cancellable but usually returned quickly and then you could use cancellation with individual reads.

When would we cancel this loop? When the app didn't read the whole body or aborted the response.

Reply:
Sounds good. Are we assuming a poorly written client? What happens when HttpResponseMessage.Dispose is called? Does this work as expected? What’s broken if we don’t support cancellation? What will hang and in what cases?

Reply:
Using SendAsync's cancellation token:

```cs
var cancelPipeRegistration = cancellationToken.Register(() =>
{
    requestPipe.Writer.Complete(new OperationCanceledException(cancellationToken));
    requestPipe.Reader.Complete(new OperationCanceledException(cancellationToken));
});
```

Reply:
Add the test and the idea isn’t bad but you can call complete while something else is writing to the PipeWriter. CancelPendingFlush is enough. Actually change that to call both CancelPendingRead and CancelPendingFlush and that should be enough

Reply:
Done